src/pybind/mgr/k8sevents/module.py

```diff
@@ -0,0 +1,1029 @@
+                                tstamp=None)
+                                )
+
+        return changes
+
+    def get_changes(self, server_map, osd_map, pool_map):
+        """Detect changes in maps between current observation and the last"""
+
+        changes = list()
+
+        changes.extend(self._check_hosts(server_map))
+        changes.extend(self._check_osds(server_map, osd_map))
+        changes.extend(self._check_pools(pool_map))
+
+        # TODO
+        # Could generate an event if a ceph daemon has moved hosts
+        # (assumes the ceph metadata host information is valid - which may not be the case)
+
+        return changes
+
+    def run(self):
+        self.mgr.log.warning("[INF] Ceph configuration watcher started")
+
+        self.server_map, self.service_map = self.fetch_servers()
+        self.pool_map = self.fetch_pools()
+        # self.mgr.log.warning("[DBG] server map {}".format(self.server_map))
+        # self.mgr.log.warning("[DBG] service map {}".format(self.service_map))
+        self.osd_map = self.fetch_osd_map(self.service_map)
+        # self.mgr.log.warning("[DBG] osd map {}".format(self.osd_map))
+
+        ctr = 0
+        
+        while True:
+
+            try:
+                start_time = time.time()
+                server_map, service_map = self.fetch_servers()
+                pool_map = self.fetch_pools()
+                osd_map = self.fetch_osd_map(service_map)
+
+                changes = self.get_changes(server_map, osd_map, pool_map)
+                if changes:
+                    self.push_events(changes)
+
+                self.osd_map = osd_map
+                self.pool_map = pool_map
+                self.server_map = server_map
+                self.service_map = service_map
+
+                checks_duration = time.time() - start_time
+                if checks_duration > self.config_check_secs:
+                    new_interval = self.config_check_secs * 2
+                    self.mgr.log.warning("[WRN] k8sevents check interval warning. "
+                                         "Current checks {}s, interval is {}. "
+                                         "Increasing interval to {}".format(int(checks_duration),
+                                                                            self.config_check_secs,
+                                                                            new_interval))
+                    self.config_check_secs = new_interval
+
+                time.sleep(self.config_check_secs)    
+                ctr += self.config_check_secs
+                if ctr%self.heartbeat_interval_secs == 0:
+                    ctr = 0
+                    self.send_hearbeat_msg()
+            except BaseException:
+                self.health = "[ERR] Unexpected {} Exception @ line {}".format(sys.exc_info()[0].__name__, 
+                                                                               sys.exc_info()[2].tb_lineno)
+                break
```

Comment:
Catching BaseException is extraordinary, e.g. [SystemExit](https://docs.python.org/3/library/exceptions.html#SystemExit). Is there a reason for catching BaseException?

Reply:
Paranoia on my part - will change to Exception